fix(http): Run fetch request outside the angular zone

JeanMeche opened the pull request:
Having the request run in the angular zone has the consequence of triggering the CD for every read of the response stream.

This commit wraps the whole `doRequest` to run outside angular with every callback on the observer being called inside the zone.

Fixes #50979.
## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [x] Bugfix

## Does this PR introduce a breaking change?

- [x] No

JeanMeche commented on packages/common/http/src/fetch.ts:
```diff
@@ -238,6 +245,24 @@ export class FetchBackend implements HttpBackend {
+  }
+
+  private observerNext(obs: Observer<HttpEvent<any>>, next: HttpEvent<any>): void {
+    this.ngZone.run(() => {
```
We could merge with the previous one and base on the object being `HttpErrorResponse` or not, calling `error` or `next`.

JeanMeche commented on packages/common/http/src/fetch.ts:
```diff
@@ -48,12 +48,18 @@ export class FetchBackend implements HttpBackend {
   // We need to bind the native fetch to its context or it will throw an "illegal invocation"
   private readonly fetchImpl =
       inject(FetchFactory, {optional: true})?.fetch ?? fetch.bind(globalThis);
+  private readonly ngZone = inject(NgZone);
```
Should we make this zoneless ready by making it `optional:true` ?

AndrewKushnir replied:
Yes, this is a good idea, but let's do it in a followup PR.

ajafff commented on packages/common/http/src/fetch.ts:
```diff
@@ -137,14 +142,16 @@ export class FetchBackend implements HttpBackend {
-          statusText: response.statusText,
-          url: getResponseUrl(response) ?? request.urlWithParams,
-        }));
+        this.ngZone.run(() => {
```
this looks redundant because `observerSend` already enters the zone

JeanMeche replied:
Of course !

ajafff commented on packages/common/http/src/fetch.ts:
```diff
@@ -48,12 +48,18 @@ export class FetchBackend implements HttpBackend {
       const aborter = new AbortController();
-      this.doRequest(request, aborter.signal, observer)
-          .then(noop, error => observer.error(new HttpErrorResponse({error})));
+
```
the current implementation in this PR always enters the zone, even for requests that should run completely outside of the Angular zone, triggering unnecessary CD:

```ts
zone.runOutsideAngular(() => http.get('/').subscribe(console.log));
```

consider checking whether the subscription happens inside the Angular zone and only use `ngZone.run()` if necessary.

JeanMeche replied:
Good point ! I added the tests to cover that !

alan-agius4 commented on packages/common/http/src/fetch.ts:
```diff
@@ -48,21 +48,34 @@ export class FetchBackend implements HttpBackend {
+
+      // We do not want to run the request in the zone
+      // Or every await will trigger a CD cycle.
+      this.ngZone.runOutsideAngular(() => {
```
Inside of running outside of zone the entire function and re-entering in parts of the inner function would it be cleaner to just `runOutsideAngular` when needed?

Ie: when reading the buffer? This would also result in less code and more code that is minifiable as class methods are not.

JeanMeche committed:
fix(http): Run fetch request out the angular zone

Having the request run in the angular zone has the consequence of triggering the CD for every read of the response stream.

This commit wraps the whole `doRequest` to run outside angular with every callback on the observer being called inside the zone.

Fixes #50979.

AndrewKushnir commented on packages/common/http/src/fetch.ts:
```diff
@@ -108,29 +109,35 @@ export class FetchBackend implements HttpBackend {
+      const isAngularZone = NgZone.isInAngularZone();
+      await this.ngZone.runOutsideAngular(async () => {
+        while (true) {
+          const {done, value} = await reader.read();
```
I have a question similar to what @alan-agius4 mentioned: can we just wrap **this call** into the `runOutsideAngular` (and `TextDecoder.decode` if we patch it in Zone.js)?

JeanMeche replied:
If you're talking about doing ` const {done, value} = await this.ngZone.runOutsideAngular(async () => reader.read());`, we'll still have that multiple await which trigger the CD.

alan-agius4 replied:
You need the await the promises inside the `this.ngZone.runOutsideAngular`.

ajafff replied:
> You need the await the promises inside the this.ngZone.runOutsideAngular.

how is that supposed to work? sure, you can await in `runOutsideAngular`, but you also need to await the resulting promise outside of the call to `runOutsideAngular` to make the loop work correctly by waiting for each chunk to be processed. this in turn causes additional change detection just like the version before this PR

AndrewKushnir reviewed (commented):
@JeanMeche thanks for the fix 👍

alxhub commented on packages/common/http/src/fetch.ts:
```diff
@@ -108,29 +109,35 @@ export class FetchBackend implements HttpBackend {
-            loaded: receivedLength,
-            partialText,
-          } as HttpDownloadProgressEvent);
+      const isAngularZone = NgZone.isInAngularZone();
```
An easier way of doing this is to capture const reqZone = Zone.current outside the closure, and then use reqZone.run() to wrap the actual emission.

JeanMeche replied:
Nice improvement thx !

AndrewKushnir commented on packages/common/http/src/fetch.ts:
```diff
@@ -108,29 +109,32 @@ export class FetchBackend implements HttpBackend {
-            partialText,
-          } as HttpDownloadProgressEvent);
+      const reqZone = Zone.current;
+      await this.ngZone.runOutsideAngular(async () => {
```
Minor proposal: let's add a comment in the code to describe why the `runOutsideAngular` is needed. Something like this:

```suggestion
      // Perform response processing outside of Angular zone to
      // ensure no excessive change detection runs are executed. 
      await this.ngZone.runOutsideAngular(async () => {
```

It'd be also great to add to that comment what exactly is triggering change detection (so it's easier to support this code in the future).

AndrewKushnir reviewed (approved):
@JeanMeche looks great, just a quick note to add a comment to document the context. Thank you.

AndrewKushnir commented:
[Presubmit](http://test/OCL:548247064:BASE:548247083:1689376452089:45f462a4).

jessicajaniuk commented:
This PR was merged into the repository by commit c5608e5ca99805af1a3a7caf4ce28a35f3a13ebf.
